### DISCUSSION TITLE
how to access script setup methods via console?

### DISCUSSION THREAD

**Discussion by tasiotas** [nuxt/nuxt#25888] [Questions]
Hi,

I defined `save()` function in the `<script setup>` of an SFC page (not component if that make a difference).

How can I access it and execute it via vanilla JS, from console for example?

I tried to find it through `window.useNuxtApp().vueApp` but I cannot find it there.
If Vue Dev Tools can do it via selecting component => $vm.setupState, than it should be possible without Dev Tools too.

Thank you

**Comment by tasiotas** [comment]
so far I found this method

```js
<script setup>
const nuxtApp = useNuxtApp()
nuxtApp.provide('save', () => save())
</script>
```

then in console I can do
```js
  const nuxtApp = window.useNuxtApp()
  nuxtApp.$save()
```

Still would be good to know how to do it without `provide`

**Comment by guibulator** [comment]
Everything defined inside script setup is private to the component. You can use `expose` to expose functions to parent component. You can read more on Vue website. https://vuejs.org/api/composition-api-setup

**Reply by tasiotas** [reply]
thanks, `expose` doesn't explain how to access it from browser's console

**Reply by guibulator** [reply]
I know, it's just explaining how to expose it in the component tree.   Can I ask why you want to access things from browser console? I never had to do this so I'm a bit curious of your use case.

**Reply by tasiotas** [reply]
sure, it is bit unusual case. 

So I have a content builder app where I simulate how the content will look on different devices (mobile, tablet, desktop).
To simulate screen size I use `iframe` to trigger correct media queries and resize content accordingly.

Now, because I have an `iframe`, can no longer communicate with its content from the parent, I need to use `postmessage` or get its content via `iframeRef.value.contentWindow.window`. So in the end I arrive at the `window` object and need to call methods from there.

**Reply by guibulator** [reply]
I see, in that case, couldn't you just expose some properties on the window object and have some watchers looking into these properties and react to it from a composable or component to these property changes? That could be a way to make your Vue app react to any external changes.

**Reply by tasiotas** [reply]
could you elaborate on "expose some properties on the window object" ? I didn't get it.

**Reply by guibulator** [reply]
Why can't you just listen to messages using `window.addEventListener("message", callback)`? So in the iframe app, you use `window.parent.postMessage` and in your parent vue app, you just watch for messages? You'll then be in a `vue` context in your parent component. You can also listen for messages the same way in your iframe component. So I'm not sure why you would need to access things from window.
There is a good gist here: [https://gist.github.com/pbojinov/8965299](https://gist.github.com/pbojinov/8965299)
